okio/src/jvmMain/kotlin/okio/JvmOkio.kt

```diff
@@ -189,6 +190,20 @@ fun Path.sink(vararg options: OpenOption): Sink =
 fun Path.source(vararg options: OpenOption): Source =
     Files.newInputStream(this, *options).source()
 
+/**
+ * Returns a sink that processes data using this [Cipher] while writing to [sink].
+ *
+ * @throws IllegalArgumentException if this isn't a block cipher.
+ */
+fun Cipher.sink(sink: BufferedSink): CipherSink = CipherSink(sink, this)
```

Comment:
Say we have the use case of writing compressed and encrypted data to a socket:

```java
BufferedSink sink = Okio.buffer(new GzipSink(new CipherSink(Okio.buffer(Okio.sink(socket)), cipher)));
// or
BufferedSink sink = Okio.buffer(new GzipSink(Okio.sink(cipher, Okio.buffer(Okio.sink(socket)))));
```

```kotlin
val sink = cipher.sink(socket.sink().buffer()).gzip().buffer()
```

**My 2 cents:**

I would submit the Java examples are easier to read than the Kotlin example which mixes fluent and non-fluent style functions. Consider the following Kotlin alternative.

```kotlin
val sink = socket.sink().cipher(cipher).gzip().buffer()
```

I think this is easier to read as it only uses fluent style functions. I also think this leads to easier discoverability as the `cipher` function is an extension of a Sink. But again, my 2 cents.

Reply:
Sure. What should we change?

Reply:
I also tend to prefer the readability of the alternative (which was I had initially suggested actually) where the extension method is on the sink/source rather than on the cipher instance. 

```kotlin
fun BufferedSink.cipher(cipher: Cipher) CipherSink = CipherSink(this, cipher)
```

In terms of discoverability, I do see why you could argue that having the extension methods on the `Cipher` class is better. It's tied to the same domain, and avoids potentially polluting the suggestions when looking through the list of methods on sink/source. 

I guess I'd prefer having code that's easy to read than that's easy to write 😄

Reply:
Ahhh! I tried by swapping parametre & receiver, creating these methods:

```
fun BufferedSink.cipher(cipher: Cipher): CipherSink
fun BufferedSource.cipher(cipher: Cipher): CipherSource
```

Unfortunately with this, calling `Buffer.cipher()` is ambiguous because both methods match.

Reply:
Oh yeah I hadn't considered that, it would have to be something like `cipherSink()`/`cipherSource()`, which is admittedly a bit verbose...

Reply:
The `gzip()` extension suffers from the same plight; you have to cast a Buffer before you can apply it. I've not found it greatly annoying in practice as I don't encounter it very often, but it is a rough edge to the gzip API, and `cipherSink()` and `cipherSource()` is an okay work-around for the cipher API.

I wish you can check the opmode of a Cipher after it has been initialized. Then you could consider the extension functions `Sink.enrypt()` and `Source.decrypt()` (similar to `inflate()`/`deflate()`).